File: distributed-job-manager/src/main/java/ru/fix/distributed/job/manager/strategy/AbstractAssignmentStrategy.java
package ru.fix.distributed.job.manager.strategy;

import ru.fix.distributed.job.manager.model.JobId;
import ru.fix.distributed.job.manager.model.WorkItem;

import java.util.Set;
import java.util.stream.Collectors;

public abstract class AbstractAssignmentStrategy implements AssignmentStrategy {

    protected Set<WorkItem> getWorkItemsByJob(JobId jobId, Set<WorkItem> workItems) {
        return workItems.stream()
                .filter(item -> item.getJobId().equals(jobId))
                .collect(Collectors.toSet());
    }
}

File: distributed-job-manager/src/main/kotlin/ru/fix/distributed/job/manager/strategy/AbstractAssignmentStrategy.kt
package ru.fix.distributed.job.manager.strategy

import ru.fix.distributed.job.manager.JobId
import ru.fix.distributed.job.manager.model.WorkItem

abstract class AbstractAssignmentStrategy : AssignmentStrategy {

    protected fun getWorkItemsByJob(jobId: JobId, workItems: Set<WorkItem>): Set<WorkItem> {
        return workItems
            .asSequence()
            .filter { item -> item.jobId == jobId }
            .toSet()
    }

    protected fun getWorkItemsByJobAsMap(workItems: Set<WorkItem>, jobs: Set<JobId>): Map<JobId, Set<WorkItem>> {
        return jobs
            .asSequence()
            .map { it to getWorkItemsByJob(it, workItems) }
            .sortedByDescending { (_, value) -> value.size }
            .toMap()
    }
}
